### PR TITLE
Fix three pre-existing bugs in JIT/compile utilities

### LINKED ISSUE #2879: Handle JIT caching for functions with closure variables
When a function passed to `@iron.jit` captures variables from outside its scope (closures), the captured value is baked into the compiled kernel on the first run. Subsequent runs with different values (if the variable is parametrized) reuses the cached kernel with the initial value.

#### Example
``` python
@pytest.mark.parametrize("add_value", [1, 2, 3])
def test_transform(add_value):
    input = iron.randint(0, 100, (1024,), dtype=np.int32, device="npu")
    output = iron.zeros_like(input)
    
    # add_value is captured from enclosing scope
    iron.jit(is_placed=False)(transform)(
        lambda a: a + add_value, input, output, tile_size=16
    )
    
    # FAILS for add_value > 1: output always reflects add_value=1
    assert np.allclose(input.numpy() + add_value, output.numpy())
```

#### Current Workaround
Caching is disabled when closures are detected:
``` python
has_closures = any(_has_closure(arg) for arg in args)
effective_use_cache = use_cache and not has_closures
```

#### Reference
[Triton](https://github.com/triton-lang/triton/pull/3762) disallows access to globals that are not tl.constexpr.

### PR DESCRIPTION
**Pull request by hunhoffe**
# Fix three pre-existing bugs in JIT/compile utilities                                                                                                    

### `jit.py` — remove stale-cache workaround for closures

`_create_function_cache_key` previously ignored closure cell contents, so any two lambdas with the same bytecode but different captured values produced the same cache key. The workaround was to disable caching entirely whenever a closure was detected (`has_closures` flag). This made the in-memory cache useless for all parametrized `@iron.jit` calls.

**Fix:** Implement `_closure_key` / `_cell_val_to_key` in `cache/utils.py` to produce a stable, value-based key component from each closure cell. Remove the `has_closures` workaround and the `_has_closure` helper. Fix the circular import of `ExternalFunction` by moving it inside `decorator` and `_filter_tensor_args`.

`_cell_val_to_key` uses a four-level priority chain:
1. **value-hash** — O(1) for types with a proper value-based `__hash__` (int, str, float, tuple, …)
2. **pickle** — SHA-256 of `pickle.dumps(val)` captures full object state (`__dict__`/`__getstate__`) for mutable objects
3. **pickle_dict** — for locally-defined classes (not picklable by name), pickle `__dict__` directly
4. **repr** — last resort

### `compile/utils.py` — `compile_external_kernel` silently swallowed missing source file

When `source_file` did not exist, the function returned silently, producing a confusing downstream linker error. **Fix:** raise `FileNotFoundError` with the missing path.

### `hash_module` — cache key collision across kernel lists

`"|".join(running_hash)` iterated over characters of a concatenated string instead of over per-kernel hash values, causing false collisions (e.g. hashes `"1"` + `"2"` collided with hash `"12"`). **Fix:** hash each kernel digest individually.

### Tests

Regression tests added to `test/python/npu-xrt/test_jit_utils.py` covering all three bugs and their edge cases (20 tests total, requires NPU hardware).

Closes #2879

### CODE REVIEW

**Review by jgmelber — APPROVED** [review]

File: python/utils/compile/cache/utils.py [hunk 1]
```diff
@@ -59,8 +117,15 @@ def _create_function_cache_key(function, args, kwargs):
...
+                    (
+                        code.co_code,
+                        code.co_consts,
+                        code.co_names,
+                        defaults,
+                        closure_vals,
+                    )
                 )
                 signature_parts.append(f"{key}_function_{func_hash}")
             else:
```
[the remark this answers is not in the document]
Comment (hunhoffe) [reply python/utils/compile/cache/utils.py:171]: Addressed

File: test/python/npu-xrt/test_jit_utils.py [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+# or raising when source_file did not exist, leaving the caller to encounter
+# a confusing downstream linker error.
+# ---------------------------------------------------------------------------
+
+
+def test_compile_external_kernel_missing_source_file_raises(npu_target_arch):
+    """FileNotFoundError must be raised when source_file does not exist.
+
+    ExternalFunction reads source_file at construction time (for hashing), so
+    the error fires in __init__ before compile_external_kernel is even called.
```
[the remark this answers is not in the document]
Comment (hunhoffe) [reply test/python/npu-xrt/test_jit_utils.py]: Addressed
